File: src/services/audit/detectors/ad/advanced/credentials/rbcd-abuse.ts
/**
 * RBCD Abuse Detector
 * Check for RBCD abuse (Resource-Based Constrained Delegation)
 */

import { ADUser } from '../../../../../../types/ad.types';
import { Finding } from '../../../../../../types/finding.types';
import { toAffectedUserEntities } from '../../../../../../utils/entity-converter';

export function detectRbcdAbuse(users: ADUser[], includeDetails: boolean): Finding {
  const affected = users.filter((u) => {
    const rbcdAttr = (u as any)['msDS-AllowedToActOnBehalfOfOtherIdentity'];
    // Check that attribute exists, is not null, not empty string, and not empty array
    return rbcdAttr && (Array.isArray(rbcdAttr) ? rbcdAttr.length > 0 : rbcdAttr !== '');
  });

  return {
    type: 'RBCD_ABUSE',
    severity: 'critical',
    category: 'advanced',
    title: 'RBCD Abuse',
    description: 'msDS-AllowedToActOnBehalfOfOtherIdentity configured. Enables privilege escalation via resource-based delegation.',
    count: affected.length,
    affectedEntities: includeDetails ? toAffectedUserEntities(affected) : undefined,
  };
}

File: internal/audit/detectors/ad/advanced/credentials/rbcd-abuse.go
package credentials

import (
	"context"

	"github.com/etcsec-com/etc-collector/internal/audit"
	"github.com/etcsec-com/etc-collector/internal/audit/helpers"
	"github.com/etcsec-com/etc-collector/pkg/types"
)

// RBCDAbuseDetector detects RBCD abuse configurations
type RBCDAbuseDetector struct {
	audit.BaseDetector
}

// NewRBCDAbuseDetector creates a new detector
func NewRBCDAbuseDetector() *RBCDAbuseDetector {
	return &RBCDAbuseDetector{
		BaseDetector: audit.NewBaseDetector("RBCD_ABUSE", audit.CategoryAdvanced),
	}
}

// Detect executes the detection
func (d *RBCDAbuseDetector) Detect(ctx context.Context, data *audit.DetectorData) []types.Finding {
	var affected []types.User

	for _, u := range data.Users {
		// Check msDS-AllowedToActOnBehalfOfOtherIdentity attribute
		if len(u.AllowedToActOnBehalfOfOtherIdentity) > 0 {
			affected = append(affected, u)
		}
	}

	finding := types.Finding{
		Type:        d.ID(),
		Severity:    types.SeverityCritical,
		Category:    string(d.Category()),
		Title:       "RBCD Abuse",
		Description: "msDS-AllowedToActOnBehalfOfOtherIdentity configured. Enables privilege escalation via resource-based delegation.",
		Count:       len(affected),
	}

	if data.IncludeDetails && len(affected) > 0 {
		finding.AffectedEntities = helpers.ToAffectedUserEntities(affected)
	}

	return []types.Finding{finding}
}

func init() {
	audit.MustRegister(NewRBCDAbuseDetector())
}
